src/random_ray/random_ray.cpp

```diff
@@ -0,0 +1,284 @@
+        settings::random_ray_distance_inactive - distance_travelled_;
+      attenuate_flux(distance_dead, false);
+
+      double distance_alive = distance - distance_dead;
+
+      // Ensure we haven't travelled past the active phase as well
+      if (distance_alive > settings::random_ray_distance_active) {
+        distance_alive = settings::random_ray_distance_active;
+        wgt() = 0.0;
+      }
+      attenuate_flux(distance_alive, true);
+
+      distance_travelled_ = distance_alive;
+    } else {
+      distance_travelled_ += distance;
+      attenuate_flux(distance, false);
+    }
+  }
+
+  // Advance particle
+  for (int j = 0; j < n_coord(); ++j) {
+    coord(j).r += distance * coord(j).u;
+  }
+}
+
+// This function forms the inner loop of the random ray transport process.
+// It is responsible for several tasks. Based on the incoming angular flux
+// of the ray and the source term in the region, the outgoing angular flux
+// is computed. The delta psi between the incoming and outgoing fluxes is
+// contributed to the estimate of the total scalar flux in the source region.
+// Additionally, the contribution of the ray path to the stochastically
+// estimated volume is also kept track of. All tasks involving writing
+// to the data for the source region are done with a lock over the entire
+// source region.  Locks are used instead of atomics as all energy groups
+// must be written, such that locking once is typically much more efficient
+// than use of many atomic operations corresponding to each energy group
+// individually (at least on CPU). Several other bookeeping tasks are also
+// performed when inside the lock.
+void RandomRay::attenuate_flux(double distance, bool is_active)
+{
+  int negroups = data::mg.num_energy_groups_;
+
+  // The number of geometric intersections is counted for reporting purposes
+  n_event()++;
+
+  // Determine source region index etc.
+  int i_cell = lowest_coord().cell;
+
+  // The source region is the spatial region index
+  int64_t source_region =
+    random_ray::source_region_offsets[i_cell] + cell_instance();
+
+  // The source element is the energy-specific region index
+  int64_t source_element = source_region * negroups;
+  int material = this->material();
+
+  // Temperature and angle indices, if using multiple temperature
+  // data sets and/or anisotropic data sets.
+  // TODO: Currently assumes we are only using single temp/single
+  // angle data.
+  const int t = 0;
+  const int a = 0;
+
+  // MOC incoming flux attenuation + source contribution/attenuation equation
+  for (int e = 0; e < negroups; e++) {
+    float Sigma_t = data::mg.macro_xs_[material].get_xs(
+      MgxsType::TOTAL, e, NULL, NULL, NULL, t, a);
+    float tau = Sigma_t * distance;
+    float exponential = cjosey_exponential(tau); // exponential = 1 - exp(-tau)
+    float new_delta_psi =
+      (angular_flux_[e] - random_ray::source[source_element + e]) * exponential;
+    delta_psi_[e] = new_delta_psi;
```

Comment:
Why do we need a separate array for `delta_psi_`? Seems like you could realize some speed benefit by keeping this as a temporary; it's just less data to move around. I get that this would mean putting the `is_active` condition check inside the energy group loop as is. That's not really a great approach.

Instead, what if you used a `constexpr if` inside the loop, with`is_active` being a template parameter? Then it compiles into two functions specialized for inactive and active mode, no branching in the performance critical loop. Locking also could be inside `constexpr if`s.

Removing the extra array will also play better with a potential GPU implementation. Since obviously this isn't something you'd want in global memory, it means more rays can fit on an SM. It would be awesome to have this some day, considering how nicely GPUs might play with random ray.

Reply:
On CPU at least, I have typically been worried that the lock contention can get to be pretty expensive between threads, so that you want to do the bare minimum of work inside the locked region to minimize the contention costs. If we moved the lock so as to encompass the actual flux attenuation computations as well (and associated cache misses when reading from the source), then I worry this might slow things down. It'd be worthwhile to test out though -- perhaps we can try it as a follow up PR with some comparative performance numbers?

Reply:
I've added this to the project as #2844 -- a number of other potential optimizations to try on this function as well. For higher group counts, it accounts for nearly the entire simulation runtime, so will be important to highly optimize.